x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/queries/SparseVectorQueryBuilder.java

```diff
@@ -0,0 +1,369 @@
+            queryRewriteContext.registerAsyncAction(
+                (client, listener) -> executeAsyncWithOrigin(
+                    client,
+                    ML_ORIGIN,
+                    CoordinatedInferenceAction.INSTANCE,
+                    inferRequest,
+                    ActionListener.wrap(inferenceResponse -> {
+
+                        if (inferenceResponse.getInferenceResults().isEmpty()) {
+                            listener.onFailure(new IllegalStateException("inference response contain no results"));
+                            return;
+                        }
+
+                        if (inferenceResponse.getInferenceResults().get(0) instanceof TextExpansionResults textExpansionResults) {
+                            textExpansionResultsSupplier.set(textExpansionResults);
+                            listener.onResponse(null);
+                        } else if (inferenceResponse.getInferenceResults().get(0) instanceof WarningInferenceResults warning) {
+                            listener.onFailure(new IllegalStateException(warning.getWarning()));
+                        } else {
+                            listener.onFailure(
+                                new IllegalStateException(
+                                    "expected a result of type ["
+                                        + TextExpansionResults.NAME
+                                        + "] received ["
+                                        + inferenceResponse.getInferenceResults().get(0).getWriteableName()
+                                        + "]. Is ["
+                                        + modelId
+                                        + "] a compatible model?"
+                                )
+                            );
+                        }
+                    }, listener::onFailure)
+                )
+            );
+
+            return new SparseVectorQueryBuilder(this, textExpansionResultsSupplier);
+
+        } else {
+            return vectorDimensionsToQuery(fieldName, vectorDimensions);
+        }
+
+    }
+
+    private QueryBuilder vectorDimensionsToQuery(String fieldName, List<VectorDimension> vectorDimensions) {
+        if (tokenPruningConfig != null) {
+            VectorDimensionsQueryBuilder vectorDimensionsQueryBuilder = new VectorDimensionsQueryBuilder(
+                fieldName,
+                vectorDimensions,
+                tokenPruningConfig
+            );
+            vectorDimensionsQueryBuilder.queryName(queryName);
+            vectorDimensionsQueryBuilder.boost(boost);
+            return vectorDimensionsQueryBuilder;
+        }
+        // Note: Weighted tokens queries were introduced in 8.13.0. To support mixed version clusters prior to 8.13.0,
+        // if no token pruning configuration is specified we fall back to a boolean query.
+        // TODO this should be updated to always use a WeightedTokensQueryBuilder once it's in all supported versions.
```

Comment:
```suggestion
        // TODO this should be updated to always use a VectorDimensionsQueryBuilder once it's in all supported versions.
```

I prefer `WeightedTokensQueryBuilder` as a name to `VectorDimensionsQueryBuilder` which feels a little generic

Reply:
I'm OK changing this back, especially since what really matters is that the `weighted_tokens` query will eventually be deprecated/removed.